Testing-the-engine.md

```diff
@@ -0,0 +1,302 @@
+
+The Mockito library is an available test dependency when writing Robolectric
+tests.
+
+### Component integration tests
+
+Component tests test the interaction of multiple embedding Java classes together
+but doesn't test all production classes end-to-end. In the Android embedding
+case, we test groups of Java classes by their function in `...ComponentTest.java`
+files that are also in the `shell/platform/android/test/io/flutter/`
+directory. C++ engine parts via JNI are not tested here.
+
+Component tests are also Robolectric JUnit tests and are invoked together with
+unit tests when running:
+
+    `testing/run_tests.py --type=java`
+
+JUnit component tests are executed during pre-submit on our CI system when
+submitting PRs to the `flutter/engine` repository.
+
+### End-to-end tests
+
+End-to-end tests exercise the entire Android embedding with the C++ engine on
+a real Android runtime in an emulator. It's an integration test ensuring that
+the engine as a whole on Android is functioning correctly.
+
+The project containing the Android end-to-end engine test is at
+https://github.com/flutter/engine/tree/master/testing/scenario_app/android.
+
+This test project is build similarly to a normal Flutter app. With the Dart code
+compiled into AOT and the Android part compiled via Gradle with a dependency
+on the prebuilt local engine. It's then installed and executed on an emulator.
+
+Unlike a normal Flutter app, the Flutter framework on the Dart side is a
+lightweight fake at https://github.com/flutter/engine/tree/master/testing/scenario_app/lib
+that implements some of the basic functionalities of `dart:ui` Window rather
+than using the real Flutter framework at `flutter/flutter`.
+
+The end-to-end test can be executed by running:
+
+    `testing/scenario_app/build_and_run_android_tests.sh`
+
+Additional end-to-end instrumented tests can be added to https://github.com/flutter/engine/tree/master/testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenarios.
+
+If supporting logic are needed for the test case, they can be added to the
+Android app under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/android/app/src/main/java/dev/flutter/scenarios
+or to the fake Flutter framework under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/lib.
+
+As best practice, favor adding unit tests if possible since instrumented tests
+are, by nature, non-hermetic, slow and flaky.
+
+End-to-end tests on Android are currently not running on CI [#55326](https://github.com/flutter/flutter/issues/55326)
+
+## Objective-C - iOS embedding
+
+If you're editing `.h` or `.mm` files in the https://github.com/flutter/engine/tree/master/shell/platform/darwin/ios
+directory, you're working on the iOS embedding which connects the core C++
+engine to the iOS SDK APIs and runtime.
+
+### XCTest unit tests
+
+For testing logic within a class at a unit level, create or add to a XCTestCase
+in the XCTest project.
```

Comment:
~~XCTest project~~ XCTest GN target

"XCTest Project" may lead people to think they are editing Apple's stuff.